### DISCUSSION TITLE
Surprising value of inf_du when warm-starting.

### DISCUSSION THREAD

**Discussion by forrestlaine** [coin-or/Ipopt#558] [Q&A] [answered]
*Moving from issues page*

I was playing around with an idea and noticed something odd. I first solve a problem and restrict the number of iterations to be 1. I then create a new solver instance, and warm-start with the output of the first solution call (1 iteration in). I initialize the primal and dual solutions, and specify the warm-start parameters as seen below. What is surprising is that at initialization of the second solve, the inf_du value does not match what it was at the termination of the first solve. Is this expected behavior? I would like to instantiate the solver for the second call at precisely the internal state it was at after 1 iteration of the first call. Apologies if this is not the correct place to ask this question. 

## First Solve
```
iter    objective    inf_pr   inf_du lg(mu)  ||d||  lg(rg) alpha_du alpha_pr  ls
   0  0.0000000e+00 9.44e+00 7.14e+00  -1.0 0.00e+00    -  0.00e+00 0.00e+00   0
   1  2.6440987e+00 9.44e+00 9.75e+12  -1.0 1.45e+12    -  1.00e+00 6.98e-14h  9
```
## Second Solve
```
iter    objective    inf_pr   inf_du lg(mu)  ||d||  lg(rg) alpha_du alpha_pr  ls
   0  2.6440987e+00 9.44e+00 7.14e+00  -1.0 0.00e+00    -  0.00e+00 0.00e+00   0
   1 -3.2218354e+02 9.27e+00 1.09e+04  -1.0 5.08e+02    -  1.97e-04 3.12e-02f  1
   2 -3.2458808e+02 9.20e+00 1.08e+04  -1.0 1.93e+01    -  6.72e-03 7.43e-03f  1
...
```
## Ipopt parameters (note I am calling the C-interface via Julia / IPOPT.jl. These options are used in both the first and second calls above. ) 
        AddIpoptStrOption(prob, "warm_start_init_point", "yes")
        AddIpoptStrOption(prob, "nlp_scaling_method", "none")
        AddIpoptNumOption(prob, "warm_start_bound_frac", 1e-19)
        AddIpoptNumOption(prob, "warm_start_bound_push", 1e-19)
        AddIpoptNumOption(prob, "warm_start_mult_bound_push", 1e-19)
        AddIpoptNumOption(prob, "warm_start_slack_bound_frac", 1e-19)
        AddIpoptNumOption(prob, "warm_start_slack_bound_push", 1e-19)

**Comment by svigerske** [comment, marked as answer]
The Julia interface builds on the C interface, which again builds on the C++ interface. So there is some complexity.   
Thus, please first check that the dual solution correctly arrives in Ipopt.

If you set print_level to 8 or higher, then from the first solve, you should get an output like
```
EXIT: Maximum Number of Iterations Exceeded.
DenseVector "final x unscaled" with 4 elements:
final x unscaled[    1]= 1.1439022675208854e+00
final x unscaled[    2]= 4.3840147530718321e+00
final x unscaled[    3]= 4.3454200666146274e+00
final x unscaled[    4]= 1.1567941245346034e+00
DenseVector "final y_c unscaled" with 1 elements:
final y_c unscaled[    1]=-4.8469599878400366e-01
DenseVector "final y_d unscaled" with 1 elements:
final y_d unscaled[    1]= 1.0650528770703656e+00
DenseVector "final z_L unscaled" with 4 elements:
final z_L unscaled[    1]= 1.3347286621844692e+01
final z_L unscaled[    2]= 7.6990390553312937e-01
final z_L unscaled[    3]= 7.7301610290806710e-01
final z_L unscaled[    4]= 1.2935615609458456e+01
DenseVector "final z_U unscaled" with 4 elements:
final z_U unscaled[    1]= 7.3385235668910154e-01
final z_U unscaled[    2]= 3.1810754011883746e-01
final z_U unscaled[    3]= 1.0000000000000009e-02
final z_U unscaled[    4]= 7.3488411359686012e-01
```
And for the second solve, you should get output starting with something like
```
DenseVector "initial x unscaled" with 4 elements:
initial x unscaled[    1]= 1.0000000000000000e+00
initial x unscaled[    2]= 5.0000000000000000e+00
initial x unscaled[    3]= 5.0000000000000000e+00
initial x unscaled[    4]= 1.0000000000000000e+00
...
```

If you do not get the duals from the first solve, maybe something needs to be done on the Julia side. There is a mention of a `NLPBlockDualStart` in their Ipopt interface:
https://github.com/jump-dev/Ipopt.jl/blob/a6b0deb7a00ce6d63c021af920f56cc273a142f6/src/MOI_wrapper.jl#L1180-L1202

If Ipopt receives the correct duals, then the following detailed Ipopt log will hopefully indicate what is happening with them next.

**Reply by forrestlaine** [reply]
Thanks for your quick response. I have checked, and indeed Ipopt receives the correct duals. As far as I can tell, none of these values are modified from the values that I pass in. I am not sure how to further diagnose the problem. 

Here is the log:
## After First Call
final x unscaled[    1]=-1.0086017571043857e-01
final x unscaled[    2]= 0.0000000000000000e+00
final x unscaled[    3]=-7.5538203573690121e-01
final x unscaled[    4]=-9.4374061485794769e+00
final x unscaled[    5]=-1.0086017571043857e-02
final x unscaled[    6]= 0.0000000000000000e+00
final x unscaled[    7]=-8.9653489520362409e-02
final x unscaled[    8]= 0.0000000000000000e+00
final x unscaled[    9]=-7.5639063749400559e-01
final x unscaled[   10]=-9.4374061485794769e+00
final x unscaled[   11]=-1.9051366523080100e-02
final x unscaled[   12]= 0.0000000000000000e+00
final x unscaled[   13]=-7.8446803330286211e-02
final x unscaled[   14]= 0.0000000000000000e+00
final x unscaled[   15]=-7.5829577414631366e-01
final x unscaled[   16]=-9.4374061485794769e+00
final x unscaled[   17]=-2.6896046856108718e-02
final x unscaled[   18]= 0.0000000000000000e+00
final x unscaled[   19]=-6.7240117140209957e-02
final x unscaled[   20]= 0.0000000000000000e+00
final x unscaled[   21]=-7.6098537883192452e-01
final x unscaled[   22]=-9.4374061485794769e+00
final x unscaled[   23]=-3.3620058570129716e-02
final x unscaled[   24]= 0.0000000000000000e+00
final x unscaled[   25]=-5.6033430950133647e-02
final x unscaled[   26]= 0.0000000000000000e+00
final x unscaled[   27]=-7.6434738468893748e-01
final x unscaled[   28]=-9.4374061485794769e+00
final x unscaled[   29]=-3.9223401665143078e-02
final x unscaled[   30]= 0.0000000000000000e+00
final x unscaled[   31]=-4.4826744760057283e-02
final x unscaled[   32]= 0.0000000000000000e+00
final x unscaled[   33]=-7.6826972485545175e-01
final x unscaled[   34]=-9.4374061485794769e+00
final x unscaled[   35]=-4.3706076141148804e-02
final x unscaled[   36]= 0.0000000000000000e+00
final x unscaled[   37]=-3.3620058569980869e-02
final x unscaled[   38]= 0.0000000000000000e+00
final x unscaled[   39]=-7.7264033246956665e-01
final x unscaled[   40]=-9.4374061485794769e+00
final x unscaled[   41]=-4.7068081998146889e-02
final x unscaled[   42]= 0.0000000000000000e+00
final x unscaled[   43]=-2.2413372379904397e-02
final x unscaled[   44]= 0.0000000000000000e+00
final x unscaled[   45]=-7.7734714066938138e-01
final x unscaled[   46]=-9.4374061485794769e+00
final x unscaled[   47]=-4.9309419236137325e-02
final x unscaled[   48]= 0.0000000000000000e+00
final x unscaled[   49]=-1.1206686189827876e-02
final x unscaled[   50]= 0.0000000000000000e+00
final x unscaled[   51]=-7.8227808259299503e-01
final x unscaled[   52]=-9.4374061485794769e+00
final x unscaled[   53]=-5.0430087855120112e-02
final x unscaled[   54]= 0.0000000000000000e+00
final x unscaled[   55]= 2.4870075940801026e-13
final x unscaled[   56]= 0.0000000000000000e+00
final x unscaled[   57]=-3.1939055641658574e-02
final x unscaled[   58]=-6.5860028681322310e-13
final x unscaled[   59]=-5.0430087855095250e-02
final x unscaled[   60]= 0.0000000000000000e+00
DenseVector "final y_c unscaled" with 40 elements:
final y_c unscaled[    1]=-2.2413372380649588e-01
final y_c unscaled[    2]=-1.0921319339868136e+01
final y_c unscaled[    3]=-2.0172035142087713e-01
final y_c unscaled[    4]= 0.0000000000000000e+00
final y_c unscaled[    5]=-2.2413372380649693e-01
final y_c unscaled[    6]=-1.0921319339868136e+01
final y_c unscaled[    7]=-1.7930697904072482e-01
final y_c unscaled[    8]= 0.0000000000000000e+00
final y_c unscaled[    9]=-2.2413372380649801e-01
final y_c unscaled[   10]=-1.0921319339868136e+01
final y_c unscaled[   11]=-1.5689360666057242e-01
final y_c unscaled[   12]= 0.0000000000000000e+00
final y_c unscaled[   13]=-2.2413372380649910e-01
final y_c unscaled[   14]=-1.0921319339868139e+01
final y_c unscaled[   15]=-1.3448023428041991e-01
final y_c unscaled[   16]= 0.0000000000000000e+00
final y_c unscaled[   17]=-2.2413372380650018e-01
final y_c unscaled[   18]=-1.0921319339868139e+01
final y_c unscaled[   19]=-1.1206686190026729e-01
final y_c unscaled[   20]= 0.0000000000000000e+00
final y_c unscaled[   21]=-2.2413372380650123e-01
final y_c unscaled[   22]=-1.0921319339868139e+01
final y_c unscaled[   23]=-8.9653489520114565e-02
final y_c unscaled[   24]= 0.0000000000000000e+00
final y_c unscaled[   25]=-2.2413372380650232e-01
final y_c unscaled[   26]=-1.0921319339868139e+01
final y_c unscaled[   27]=-6.7240117139961739e-02
final y_c unscaled[   28]= 0.0000000000000000e+00
final y_c unscaled[   29]=-2.2413372380650340e-01
final y_c unscaled[   30]=-1.0921319339868139e+01
final y_c unscaled[   31]=-4.4826744759808794e-02
final y_c unscaled[   32]= 0.0000000000000000e+00
final y_c unscaled[   33]=-2.2413372380650448e-01
final y_c unscaled[   34]=-1.0921319339868141e+01
final y_c unscaled[   35]=-2.2413372379655752e-02
final y_c unscaled[   36]= 0.0000000000000000e+00
final y_c unscaled[   37]=-2.2413372380650556e-01
final y_c unscaled[   38]=-1.0921319339868141e+01
final y_c unscaled[   39]= 4.9740151881602051e-13
final y_c unscaled[   40]= 0.0000000000000000e+00
DenseVector "final y_d unscaled" with 30 elements:
final y_d unscaled[    1]=-3.5623576386725305e-14
final y_d unscaled[    2]=-2.9222405056797098e-14
final y_d unscaled[    3]=-8.7213107308889318e-16
final y_d unscaled[    4]=-3.5623577411095206e-14
final y_d unscaled[    5]=-2.9222419558750218e-14
final y_d unscaled[    6]=-8.7201900622713280e-16
final y_d unscaled[    7]=-3.5623579346016156e-14
final y_d unscaled[    8]=-2.9222446951328344e-14
final y_d unscaled[    9]=-8.7192094772309243e-16
final y_d unscaled[   10]=-3.5623582077669240e-14
final y_d unscaled[   11]=-2.9222485623203348e-14
final y_d unscaled[   12]=-8.7183689757677219e-16
final y_d unscaled[   13]=-3.5623585492235613e-14
final y_d unscaled[   14]=-2.9222533963047097e-14
final y_d unscaled[   15]=-8.7176685578817227e-16
final y_d unscaled[   16]=-3.5623589475896372e-14
final y_d unscaled[   17]=-2.9222590359531475e-14
final y_d unscaled[   18]=-8.7171082235729207e-16
final y_d unscaled[   19]=-3.5623593914832646e-14
final y_d unscaled[   20]=-2.9222653201328350e-14
final y_d unscaled[   21]=-8.7166879728413210e-16
final y_d unscaled[   22]=-3.5623598695225564e-14
final y_d unscaled[   23]=-2.9222720877109598e-14
final y_d unscaled[   24]=-8.7164078056869225e-16
final y_d unscaled[   25]=-3.5623603703256224e-14
final y_d unscaled[   26]=-2.9222791775547101e-14
final y_d unscaled[   27]=-8.7162677221097262e-16
final y_d unscaled[   28]=-6.9786052242528115e+01
final y_d unscaled[   29]=-6.9897397378223928e-15
final y_d unscaled[   30]=-8.7162677221097282e-16

## At second call
initial x unscaled[    1]=-1.0086017571043857e-01
initial x unscaled[    2]= 0.0000000000000000e+00
initial x unscaled[    3]=-7.5538203573690121e-01
initial x unscaled[    4]=-9.4374061485794769e+00
initial x unscaled[    5]=-1.0086017571043857e-02
initial x unscaled[    6]= 0.0000000000000000e+00
initial x unscaled[    7]=-8.9653489520362409e-02
initial x unscaled[    8]= 0.0000000000000000e+00
initial x unscaled[    9]=-7.5639063749400559e-01
initial x unscaled[   10]=-9.4374061485794769e+00
initial x unscaled[   11]=-1.9051366523080100e-02
initial x unscaled[   12]= 0.0000000000000000e+00
initial x unscaled[   13]=-7.8446803330286211e-02
initial x unscaled[   14]= 0.0000000000000000e+00
initial x unscaled[   15]=-7.5829577414631366e-01
initial x unscaled[   16]=-9.4374061485794769e+00
initial x unscaled[   17]=-2.6896046856108718e-02
initial x unscaled[   18]= 0.0000000000000000e+00
initial x unscaled[   19]=-6.7240117140209957e-02
initial x unscaled[   20]= 0.0000000000000000e+00
initial x unscaled[   21]=-7.6098537883192452e-01
initial x unscaled[   22]=-9.4374061485794769e+00
initial x unscaled[   23]=-3.3620058570129716e-02
initial x unscaled[   24]= 0.0000000000000000e+00
initial x unscaled[   25]=-5.6033430950133647e-02
initial x unscaled[   26]= 0.0000000000000000e+00
initial x unscaled[   27]=-7.6434738468893748e-01
initial x unscaled[   28]=-9.4374061485794769e+00
initial x unscaled[   29]=-3.9223401665143078e-02
initial x unscaled[   30]= 0.0000000000000000e+00
initial x unscaled[   31]=-4.4826744760057283e-02
initial x unscaled[   32]= 0.0000000000000000e+00
initial x unscaled[   33]=-7.6826972485545175e-01
initial x unscaled[   34]=-9.4374061485794769e+00
initial x unscaled[   35]=-4.3706076141148804e-02
initial x unscaled[   36]= 0.0000000000000000e+00
initial x unscaled[   37]=-3.3620058569980869e-02
initial x unscaled[   38]= 0.0000000000000000e+00
initial x unscaled[   39]=-7.7264033246956665e-01
initial x unscaled[   40]=-9.4374061485794769e+00
initial x unscaled[   41]=-4.7068081998146889e-02
initial x unscaled[   42]= 0.0000000000000000e+00
initial x unscaled[   43]=-2.2413372379904397e-02
initial x unscaled[   44]= 0.0000000000000000e+00
initial x unscaled[   45]=-7.7734714066938138e-01
initial x unscaled[   46]=-9.4374061485794769e+00
initial x unscaled[   47]=-4.9309419236137325e-02
initial x unscaled[   48]= 0.0000000000000000e+00
initial x unscaled[   49]=-1.1206686189827876e-02
initial x unscaled[   50]= 0.0000000000000000e+00
initial x unscaled[   51]=-7.8227808259299503e-01
initial x unscaled[   52]=-9.4374061485794769e+00
initial x unscaled[   53]=-5.0430087855120112e-02
initial x unscaled[   54]= 0.0000000000000000e+00
initial x unscaled[   55]= 2.4870075940801026e-13
initial x unscaled[   56]= 0.0000000000000000e+00
initial x unscaled[   57]=-3.1939055641658574e-02
initial x unscaled[   58]=-6.5860028681322310e-13
initial x unscaled[   59]=-5.0430087855095250e-02
initial x unscaled[   60]= 0.0000000000000000e+00
DenseVector "initial y_c unscaled" with 40 elements:
initial y_c unscaled[    1]=-2.2413372380649588e-01
initial y_c unscaled[    2]=-1.0921319339868136e+01
initial y_c unscaled[    3]=-2.0172035142087713e-01
initial y_c unscaled[    4]= 0.0000000000000000e+00
initial y_c unscaled[    5]=-2.2413372380649693e-01
initial y_c unscaled[    6]=-1.0921319339868136e+01
initial y_c unscaled[    7]=-1.7930697904072482e-01
initial y_c unscaled[    8]= 0.0000000000000000e+00
initial y_c unscaled[    9]=-2.2413372380649801e-01
initial y_c unscaled[   10]=-1.0921319339868136e+01
initial y_c unscaled[   11]=-1.5689360666057242e-01
initial y_c unscaled[   12]= 0.0000000000000000e+00
initial y_c unscaled[   13]=-2.2413372380649910e-01
initial y_c unscaled[   14]=-1.0921319339868139e+01
initial y_c unscaled[   15]=-1.3448023428041991e-01
initial y_c unscaled[   16]= 0.0000000000000000e+00
initial y_c unscaled[   17]=-2.2413372380650018e-01
initial y_c unscaled[   18]=-1.0921319339868139e+01
initial y_c unscaled[   19]=-1.1206686190026729e-01
initial y_c unscaled[   20]= 0.0000000000000000e+00
initial y_c unscaled[   21]=-2.2413372380650123e-01
initial y_c unscaled[   22]=-1.0921319339868139e+01
initial y_c unscaled[   23]=-8.9653489520114565e-02
initial y_c unscaled[   24]= 0.0000000000000000e+00
initial y_c unscaled[   25]=-2.2413372380650232e-01
initial y_c unscaled[   26]=-1.0921319339868139e+01
initial y_c unscaled[   27]=-6.7240117139961739e-02
initial y_c unscaled[   28]= 0.0000000000000000e+00
initial y_c unscaled[   29]=-2.2413372380650340e-01
initial y_c unscaled[   30]=-1.0921319339868139e+01
initial y_c unscaled[   31]=-4.4826744759808794e-02
initial y_c unscaled[   32]= 0.0000000000000000e+00
initial y_c unscaled[   33]=-2.2413372380650448e-01
initial y_c unscaled[   34]=-1.0921319339868141e+01
initial y_c unscaled[   35]=-2.2413372379655752e-02
initial y_c unscaled[   36]= 0.0000000000000000e+00
initial y_c unscaled[   37]=-2.2413372380650556e-01
initial y_c unscaled[   38]=-1.0921319339868141e+01
initial y_c unscaled[   39]= 4.9740151881602051e-13
initial y_c unscaled[   40]= 0.0000000000000000e+00
DenseVector "initial y_d unscaled" with 30 elements:
initial y_d unscaled[    1]=-3.5623576386725305e-14
initial y_d unscaled[    2]=-2.9222405056797098e-14
initial y_d unscaled[    3]=-8.7213107308889318e-16
initial y_d unscaled[    4]=-3.5623577411095206e-14
initial y_d unscaled[    5]=-2.9222419558750218e-14
initial y_d unscaled[    6]=-8.7201900622713280e-16
initial y_d unscaled[    7]=-3.5623579346016156e-14
initial y_d unscaled[    8]=-2.9222446951328344e-14
initial y_d unscaled[    9]=-8.7192094772309243e-16
initial y_d unscaled[   10]=-3.5623582077669240e-14
initial y_d unscaled[   11]=-2.9222485623203348e-14
initial y_d unscaled[   12]=-8.7183689757677219e-16
initial y_d unscaled[   13]=-3.5623585492235613e-14
initial y_d unscaled[   14]=-2.9222533963047097e-14
initial y_d unscaled[   15]=-8.7176685578817227e-16
initial y_d unscaled[   16]=-3.5623589475896372e-14
initial y_d unscaled[   17]=-2.9222590359531475e-14
initial y_d unscaled[   18]=-8.7171082235729207e-16
initial y_d unscaled[   19]=-3.5623593914832646e-14
initial y_d unscaled[   20]=-2.9222653201328350e-14
initial y_d unscaled[   21]=-8.7166879728413210e-16
initial y_d unscaled[   22]=-3.5623598695225564e-14
initial y_d unscaled[   23]=-2.9222720877109598e-14
initial y_d unscaled[   24]=-8.7164078056869225e-16
initial y_d unscaled[   25]=-3.5623603703256224e-14
initial y_d unscaled[   26]=-2.9222791775547101e-14
initial y_d unscaled[   27]=-8.7162677221097262e-16
initial y_d unscaled[   28]=-6.9786052242528115e+01
initial y_d unscaled[   29]=-6.9897397378223928e-15
initial y_d unscaled[   30]=-8.7162677221097282e-16


initial y_c[    1]=-2.2413372380649588e-01
initial y_c[    2]=-1.0921319339868136e+01
initial y_c[    3]=-2.0172035142087713e-01
initial y_c[    4]= 0.0000000000000000e+00
initial y_c[    5]=-2.2413372380649693e-01
initial y_c[    6]=-1.0921319339868136e+01
initial y_c[    7]=-1.7930697904072482e-01
initial y_c[    8]= 0.0000000000000000e+00
initial y_c[    9]=-2.2413372380649801e-01
initial y_c[   10]=-1.0921319339868136e+01
initial y_c[   11]=-1.5689360666057242e-01
initial y_c[   12]= 0.0000000000000000e+00
initial y_c[   13]=-2.2413372380649910e-01
initial y_c[   14]=-1.0921319339868139e+01
initial y_c[   15]=-1.3448023428041991e-01
initial y_c[   16]= 0.0000000000000000e+00
initial y_c[   17]=-2.2413372380650018e-01
initial y_c[   18]=-1.0921319339868139e+01
initial y_c[   19]=-1.1206686190026729e-01
initial y_c[   20]= 0.0000000000000000e+00
initial y_c[   21]=-2.2413372380650123e-01
initial y_c[   22]=-1.0921319339868139e+01
initial y_c[   23]=-8.9653489520114565e-02
initial y_c[   24]= 0.0000000000000000e+00
initial y_c[   25]=-2.2413372380650232e-01
initial y_c[   26]=-1.0921319339868139e+01
initial y_c[   27]=-6.7240117139961739e-02
initial y_c[   28]= 0.0000000000000000e+00
initial y_c[   29]=-2.2413372380650340e-01
initial y_c[   30]=-1.0921319339868139e+01
initial y_c[   31]=-4.4826744759808794e-02
initial y_c[   32]= 0.0000000000000000e+00
initial y_c[   33]=-2.2413372380650448e-01
initial y_c[   34]=-1.0921319339868141e+01
initial y_c[   35]=-2.2413372379655752e-02
initial y_c[   36]= 0.0000000000000000e+00
initial y_c[   37]=-2.2413372380650556e-01
initial y_c[   38]=-1.0921319339868141e+01
initial y_c[   39]= 4.9740151881602051e-13
initial y_c[   40]= 0.0000000000000000e+00
DenseVector "initial y_d" with 30 elements:
initial y_d[    1]=-3.5623576386725305e-14
initial y_d[    2]=-2.9222405056797098e-14
initial y_d[    3]=-8.7213107308889318e-16
initial y_d[    4]=-3.5623577411095206e-14
initial y_d[    5]=-2.9222419558750218e-14
initial y_d[    6]=-8.7201900622713280e-16
initial y_d[    7]=-3.5623579346016156e-14
initial y_d[    8]=-2.9222446951328344e-14
initial y_d[    9]=-8.7192094772309243e-16
initial y_d[   10]=-3.5623582077669240e-14
initial y_d[   11]=-2.9222485623203348e-14
initial y_d[   12]=-8.7183689757677219e-16
initial y_d[   13]=-3.5623585492235613e-14
initial y_d[   14]=-2.9222533963047097e-14
initial y_d[   15]=-8.7176685578817227e-16
initial y_d[   16]=-3.5623589475896372e-14
initial y_d[   17]=-2.9222590359531475e-14
initial y_d[   18]=-8.7171082235729207e-16
initial y_d[   19]=-3.5623593914832646e-14
initial y_d[   20]=-2.9222653201328350e-14
initial y_d[   21]=-8.7166879728413210e-16
initial y_d[   22]=-3.5623598695225564e-14
initial y_d[   23]=-2.9222720877109598e-14
initial y_d[   24]=-8.7164078056869225e-16
initial y_d[   25]=-3.5623603703256224e-14
initial y_d[   26]=-2.9222791775547101e-14
initial y_d[   27]=-8.7162677221097262e-16
initial y_d[   28]=-6.9786052242528115e+01
initial y_d[   29]=-6.9897397378223928e-15
initial y_d[   30]=-8.7162677221097282e-16

**Reply by forrestlaine** [reply]
I think I have located the issue, which is that the values of v_L and v_U are not passed in. However, it appears v_U is identically, likely because my upper bounds are strictly infinite. Can I specify the values of v_L and v_U when warm-starting?

**Reply by svigerske** [reply]
If I remember right, `v_L` and `v_U` are the dual multipliers for the slack variables that were introduced for inequality constraints. So if you have `lhs <= g(x) <= rhs` and assume lhs != rhs, then this becomes `g(x) - s = 0`, `lhs <= s <= rhs`. The values you pass in for lambda are used to initialize the dual values for `g(x) - s = 0` - they are called `y_d` (because, internally, `d(x) := g(x)-s`). But the normal starting-point callbacks do not allow to initialize `v_L` and `v_U`. Instead, `v_L` and `v_U` are derived from `y_d` (i.e., lambda), as something like `v_L = max(1e-19,-y_d)`, `v_U=max(1e-19,y_d)`:
https://github.com/coin-or/Ipopt/blob/a62679dd2cd292a9e3aa64ad05df37f5adfdece5/src/Algorithm/IpWarmStartIterateInitializer.cpp#L241-L252    
It made sense to omit `v_L` and `v_U` from `TNLP::get_starting_point()`, because the TNLP does not have these slack variables, so also no dual multipliers for their bounds. Further, the default initialization or `v_L` and `v_U` is done such corresponding KKT conditions are satisfied. So, even though you cannot reproduce the complete iterate vector of Ipopt's internal NLP, and thus may see a different dual infeasibility (`inf_du`, which is for the internal NLP), it seems sound what Ipopt is doing here.

There is another method to pass in a warm start iterate. This one includes `v_L` and `v_U`, but expects everything in terms of the internal NLP, is marked as "only for experts", and is only available in the C++ interface: https://coin-or.github.io/Ipopt/classIpopt_1_1TNLP.html#a273b8be28d2a0a805489cd8903aac1c6  (I think Bonmin is using this.)
Thus, from C interface, and thus from the Julia interface, you cannot specify the values for v_L and v_U.

I guess there are several options on how to proceed:
- You accept that you cannot restart Ipopt at exactly the same iterate. Even if one could set the complete iterate, since also the value of the Barrier parameter mu is not that easy to control, I don't think that there is a way to stop and restart Ipopt and have it look as if it was never interrupted.
- You reformulate your problem to avoid inequality constraints. That is, you introduce the slack variables on your end. The `v_L` and `v_U` will then be part of `z_L` and `z_U`.
- One extends the `get_starting_point()` functions of the Ipopt interfaces to allow specifying `v_L` and `v_U` in some way. However, as written above, it doesn't look right to allow users to specify dual values for bounds on slack variable which the user does not see. Also I wouldn't want to change existing API functions, so would need to add additional ones, which will look a bit awkward.
- One extends the C and Julia APIs to give access to `TNLP::get_warm_start_iterate()`. However, that would also require to make it possible to retrieve the `IteratesVector` at the end of your first solve.
- You switch to use the C++ API directly.

**Reply by forrestlaine** [reply]
Thank you very much for such a helpful and detailed response. I will have to ponder these options to see which makes most sense for me. I will probably work with the C++ interface for the time being to make sure that passing in these values is worth the effort. 

Again, thank you for your help!
